ComponentKit/Utilities/CKComponentAction.h

```diff
@@ -11,44 +11,177 @@
+    _CKTypedComponentDebugCheckComponentScope(scope, selector, typeEncodings);
+#endif
+  }
+  
+  /** Legacy constructor for raw selector actions. Traverse up the mount responder chain. */
+  CKTypedComponentAction(SEL selector) : _internal(CKTypedComponentActionVariantRawSelector, nil, nil, selector) { };
+
+  /** We support promotion from actions that take no arguments. */
+  template <typename T1, typename... Ts>
+  CKTypedComponentAction<T1, Ts...>(const CKTypedComponentAction<> &action) : _internal(action._internal) { };
 
-typedef NS_ENUM(NSUInteger, CKComponentActionSendBehavior) {
-  /** Starts searching at the sender's next responder. Usually this is what you want to prevent infinite loops. */
-  CKComponentActionSendBehaviorStartAtSenderNextResponder,
-  /** If the sender itself responds to the action, invoke the action on the sender. */
-  CKComponentActionSendBehaviorStartAtSender,
+  /** Const copy constructor to allow for block capture of the struct. */
+  CKTypedComponentAction<T...>(const CKTypedComponentAction<T...> &action) : _internal(action._internal) { };
+
+  /**
+   We allow demotion from actions with types to untyped actions, but only when explicit. This means arguments to the
+   method specified here will have nil values at runtime. Used for interoperation with older API's.
+   */
+  template<typename... Ts>
+  explicit CKTypedComponentAction<>(const CKTypedComponentAction<Ts...> &action) : _internal(action._internal) { };
+
+  /** Allows conversion from NULL actions. */
+  CKTypedComponentAction(int s) : _internal({}) {};
+  CKTypedComponentAction(long s) : _internal({}) {};
+  CKTypedComponentAction(std::nullptr_t n) : _internal({}) {};
+  
+  explicit operator bool() const { return bool(_internal); };
+  bool operator==(const CKTypedComponentAction& rhs) const { return _internal == rhs._internal; }
+  
+  SEL selector() const { return _internal.selector(); };
+  
+  void send(CKComponent *sender, T... args) const
+  { this->send(sender, _internal.defaultBehavior(), args...); }
+  void send(CKComponent *sender, CKComponentActionSendBehavior behavior, T... args) const
+  {
+    if (!_internal) {
+      return;
+    }
+    const id target = _internal.initialTarget(sender);
+    const id responder = behavior == CKComponentActionSendBehaviorStartAtSender ? target : [target nextResponder];
+    CKComponentActionSendResponderChain(_internal.selector(), responder, sender, args...);
+  }
+  
+  /** Allows you to get a block that sends the action when executed. */
+  typedef void (^CKTypedComponentActionCurriedSenderExecutionBlock)(T... args);
+  CKTypedComponentActionCurriedSenderExecutionBlock curriedSenderBlock(id sender, CKComponentActionSendBehavior behavior) const
+  {
+    __weak id weakSender = sender;
+    CKTypedComponentAction<T...> copy {*this};
+    return ^(T... args) {
+      id strongSender = weakSender;
+      copy.send(strongSender, behavior, args...);
+    };
+  }
+
+  CKTypedComponentActionValue _internal;
```

Comment:
I couldn't make the values private without explicit specializations/friends, so I opted for this guy, which prevents raw access to the members.